File: core/src/ru/deadsoftware/cavedroid/game/mobs/Player.java
package ru.deadsoftware.cavedroid.game.mobs;

import com.badlogic.gdx.graphics.g2d.Sprite;
import com.badlogic.gdx.graphics.g2d.SpriteBatch;
import com.badlogic.gdx.math.MathUtils;
import com.badlogic.gdx.math.Vector2;
import ru.deadsoftware.cavedroid.game.GameItemsHolder;
import ru.deadsoftware.cavedroid.game.model.block.Block;
import ru.deadsoftware.cavedroid.game.model.item.InventoryItem;
import ru.deadsoftware.cavedroid.game.model.item.Item;
import ru.deadsoftware.cavedroid.game.objects.Drop;
import ru.deadsoftware.cavedroid.game.world.GameWorld;
import ru.deadsoftware.cavedroid.misc.Assets;
import ru.deadsoftware.cavedroid.misc.utils.SpriteOrigin;
import ru.deadsoftware.cavedroid.misc.utils.SpriteUtilsKt;

import javax.annotation.CheckForNull;

public class Player extends Mob {

    private static final float SPEED = 69.072f;
    private static final float JUMP_VELOCITY = -133.332f;
    private static final int MAX_HEALTH = 20;

    private boolean hitting = false, hittingWithDamage = false;
    private float hitAnim = 0f;
    private float hitAnimDelta = ANIMATION_SPEED;

    public final InventoryItem[] inventory;
    public int slot;
    public int gameMode;
    public boolean swim;
    public float headRotation = 0f;

    public float blockDamage = 0f;
    public int cursorX = 0;
    public int cursorY = 0;

    public ControlMode controlMode = ControlMode.WALK;

    public enum ControlMode {
        WALK,
        CURSOR
    }

    public Player(GameItemsHolder gameItemsHolder) {
        super(0, 0, 4, 30, randomDir(), Type.MOB, MAX_HEALTH);
        inventory = new InventoryItem[36];
        for (int i = 0; i < inventory.length; i++) {
            inventory[i] = gameItemsHolder.getFallbackItem().toInventoryItem();
        }
        swim = false;
    }

    public void initInventory(GameItemsHolder gameItemsHolder) {
        for (InventoryItem invItem : inventory) {
            invItem.init(gameItemsHolder);
        }
    }

    @CheckForNull
    public Item inventory(int i) {
        return inventory[i].getItem();
    }
    
    public void respawn(GameWorld gameWorld, GameItemsHolder itemsHolder) {
        Vector2 pos = getSpawnPoint(gameWorld, itemsHolder);
        this.x = pos.x;
        this.y = pos.y;
        mVelocity.setZero();
        mDead = false;
        heal(MAX_HEALTH);
    }

    public void decreaseCurrentItemCount(GameItemsHolder gameItemsHolder) {
        if (gameMode == 1) {
            return;
        }
        getCurrentItem().setAmount(getCurrentItem().getAmount() - 1);
        if (getCurrentItem().getAmount() <= 0) {
            setCurrentInventorySlotItem(gameItemsHolder.getFallbackItem());
        }
    }

    public InventoryItem getCurrentItem() {
        return inventory[slot];
    }

    public void pickUpDrop(Drop drop) {
        for (InventoryItem invItem : inventory) {
            if (!invItem.getItem().isTool()
                    && invItem.getItem() == drop.getItem()
                    && invItem.getAmount() < invItem.getItem().getParams().getMaxStack()) {
                invItem.setAmount(invItem.getAmount() + 1);
                drop.setPickedUp(true);
                return;
            }
        }

        for (int i = 0; i < inventory.length; i++) {
            if (inventory(i) == null || inventory(i).getParams().getKey().equals(GameItemsHolder.FALLBACK_ITEM_KEY)) {
                inventory[i] = drop.getItem().toInventoryItem();
                drop.setPickedUp(true);
                break;
            }
        }
    }

    private Vector2 getSpawnPoint(GameWorld gameWorld, GameItemsHolder itemsHolder) {
        int y;
        for (y = 0; y < gameWorld.getHeight(); y++) {
            if (y == gameWorld.getHeight() - 1) {
                y = 60;
                gameWorld.setForeMap(0, y, itemsHolder.getBlock("grass"));
                break;
            }
            if (gameWorld.hasForeAt(0, y) && gameWorld.getForeMap(0, y).hasCollision()) {
                break;
            }
        }
        return new Vector2(8 - getWidth() / 2, (float) y * 16 - getHeight());
    }

    public void setDir(Direction dir) {
        if (dir != getDirection()) {
            switchDir();
        }
    }

    public void setCurrentInventorySlotItem(Item item) {
        inventory[slot] = item.toInventoryItem();
    }

    @Override
    public float getSpeed() {
        return SPEED;
    }

    @Override
    public void jump() {
        mVelocity.y = JUMP_VELOCITY;
    }

    private void hitBlock(GameWorld gameWorld, GameItemsHolder gameItemsHolder) {
        if (!hitting || !hittingWithDamage) {
            return;
        }

        final Block foregroundBlock = gameWorld.getForeMap(cursorX, cursorY);
        final Block backgroundBlock = gameWorld.getBackMap(cursorX, cursorY);

        if ((!foregroundBlock.isNone() && foregroundBlock.getParams().getHitPoints() >= 0) ||
                (foregroundBlock.isNone() && !backgroundBlock.isNone() && backgroundBlock.getParams().getHitPoints() >= 0)) {
            if (gameMode == 0) {
                if (!foregroundBlock.isNone() && blockDamage >= foregroundBlock.getParams().getHitPoints()) {
                    gameWorld.destroyForeMap(cursorX, cursorY);
                    blockDamage = 0;
                } else if (!backgroundBlock.isNone() && blockDamage >= backgroundBlock.getParams().getHitPoints()) {
                    gameWorld.destroyBackMap(cursorX, cursorY);
                    blockDamage = 0;
                }
            } else {
                if (!foregroundBlock.isNone()) {
                    gameWorld.placeToForeground(cursorX, cursorY, gameItemsHolder.getFallbackBlock());
                } else if (!backgroundBlock.isNone()) {
                    gameWorld.placeToBackground(cursorX, cursorY, gameItemsHolder.getFallbackBlock());
                }
                stopHitting();
            }
        } else {
            stopHitting();
        }
    }

    @Override
    public void ai(GameWorld gameWorld, GameItemsHolder gameItemsHolder, float delta) {
        updateAnimation(delta);
        hitBlock(gameWorld, gameItemsHolder);

        if (gameMode == 1) {
            return;
        }

        final Block foregroundBlock = gameWorld.getForeMap(cursorX, cursorY);
        final Block backgroundBlock = gameWorld.getBackMap(cursorX, cursorY);
        @CheckForNull final Block target;

        if (!foregroundBlock.isNone() && foregroundBlock.getParams().getHitPoints() >= 0) {
            target = foregroundBlock;
        } else if (!backgroundBlock.isNone() && backgroundBlock.getParams().getHitPoints() >= 0) {
            target = backgroundBlock;
        } else {
            target = null;
        }

        final boolean canHitBlock = target != null;

        float multiplier = 1f;
        final Item currentItem = inventory[slot].getItem();
        if (currentItem instanceof Item.Tool && canHitBlock) {
            if (target.getParams().getToolType() == currentItem.getClass()
                    && ((Item.Tool)currentItem).getLevel() >= target.getParams().getToolLevel()) {
                multiplier = 2f * ((Item.Tool)currentItem).getLevel();
            }
            multiplier *= ((Item.Tool)currentItem).getBlockDamageMultiplier();
        }

        if (hitting && hittingWithDamage && canHitBlock) {
            blockDamage += 60f * delta * multiplier;
        } else {
            blockDamage = 0f;
        }
    }

    @Override
    public void changeDir() {
    }

    @Override
    public void damage(int damage) {
        if (gameMode == 1) {
            return;
        }

        if (damage > 0) {
            getVelocity().y += JUMP_VELOCITY / 3f;
        }

        super.damage(damage);
    }

    @Override
    public void heal(int heal) {
        if (gameMode == 1) {
            return;
        }
        super.heal(heal);
    }

    private void drawItem(SpriteBatch spriteBatch, float x, float y, float anim) {
        final Item item = inventory(slot);

        if (item == null || item.isNone()) {
            return;
        }

        final Sprite sprite = item.getSprite();
        final boolean smallSprite = !item.isTool() || item.isShears();

        if (smallSprite) {
            sprite.setSize(Drop.DROP_SIZE, Drop.DROP_SIZE);
        }

        final float handLength = Assets.playerSprite[0][2].getHeight();

        final SpriteOrigin spriteOrigin = item.getParams().getInHandSpriteOrigin();
        final int handMultiplier = -getDirection().getBasis();
        final float xOffset = (-1 + getDirection().getIndex()) * sprite.getWidth() + 4 + handMultiplier * (sprite.getWidth() * spriteOrigin.getX());
        final float yOffset = !smallSprite ? -sprite.getHeight() / 2 : 0;

        float rotate = anim + 30;

        if (item.isTool()) {
            sprite.rotate90(looksLeft());
        }

        final float itemX = x + handLength * MathUtils.sin(handMultiplier * anim * MathUtils.degRad) + xOffset;
        final float itemY = y + handLength * MathUtils.cos(handMultiplier * anim * MathUtils.degRad) + yOffset;

        if (looksLeft()) {
            sprite.setFlip(!item.isTool(), sprite.isFlipY());
            SpriteUtilsKt.applyOrigin(sprite, spriteOrigin.getFlipped(true, false));
        } else {
            sprite.setFlip(item.isTool(), sprite.isFlipY());
            SpriteUtilsKt.applyOrigin(sprite, spriteOrigin);
        }

        SpriteUtilsKt.drawSprite(spriteBatch, sprite, itemX, itemY, -handMultiplier * rotate);

        // dont forget to reset
        sprite.setFlip(false, sprite.isFlipY());
        sprite.setRotation(0);
        sprite.setOriginCenter();
        if (item.isTool()) {
            sprite.rotate90(looksRight());
        }
    }

    public void startHitting(boolean withDamage) {
        if (hitting) {
            return;
        }

        hitting = true;
        hittingWithDamage = withDamage;
        hitAnim = 90f;
        hitAnimDelta = ANIMATION_SPEED;
    }

    public void startHitting() {
        startHitting(true);
    }

    public void stopHitting() {
        blockDamage = 0f;
        hitting = false;
    }

    private float getRightHandAnim(float delta) {
        hitAnim -= hitAnimDelta * delta;

        if (hitAnim < 30f || hitAnim > 90f) {
            if (hitting) {
                hitAnim = MathUtils.clamp(hitAnim, 30f, 90f);
                hitAnimDelta = -hitAnimDelta;
            } else  {
                hitAnimDelta = ANIMATION_SPEED;
            }
        }

        if (!hitting) {
            if (hitAnim < hitAnimDelta * delta) {
                hitAnim = 0;
                hitAnimDelta = 0;
                return -mAnim;
            }
        }

        return hitAnim;
    }

    @Override
    public void draw(SpriteBatch spriteBatch, float x, float y, float delta) {
        final Sprite backHand = Assets.playerSprite[1][2];
        final Sprite backLeg = Assets.playerSprite[1][3];
        final Sprite frontLeg = Assets.playerSprite[0][3];
        final Sprite head = Assets.playerSprite[getDirection().getIndex()][0];
        final Sprite body = Assets.playerSprite[getDirection().getIndex()][1];
        final Sprite frontHand = Assets.playerSprite[0][2];

        float backHandAnim, frontHandAnim;

        final float rightHandAnim = getRightHandAnim(delta);

        if (looksLeft()) {
            backHandAnim = rightHandAnim;
            frontHandAnim = mAnim;
        } else {
            backHandAnim = -mAnim;
            frontHandAnim = -rightHandAnim;
        }

        SpriteUtilsKt.drawSprite(spriteBatch, backHand, x + 2, y + 8, backHandAnim);

        if (looksLeft()) {
            drawItem(spriteBatch, x, y, -backHandAnim);
        }

        SpriteUtilsKt.drawSprite(spriteBatch, backLeg, x + 2, y + 20, mAnim);
        SpriteUtilsKt.drawSprite(spriteBatch, frontLeg, x + 2, y + 20, -mAnim);
        SpriteUtilsKt.drawSprite(spriteBatch, head, x, y, headRotation);
        SpriteUtilsKt.drawSprite(spriteBatch, body, x + 2, y + 8);

        if (looksRight()) {
            drawItem(spriteBatch, x, y, frontHandAnim);
        }

        SpriteUtilsKt.drawSprite(spriteBatch, frontHand, x + 2, y + 8, frontHandAnim);
    }

}

File: core/entity/mob/src/main/kotlin/ru/fredboy/cavedroid/entity/mob/model/Player.kt
package ru.fredboy.cavedroid.entity.mob.model

import box2dLight.Light
import box2dLight.PointLight
import com.badlogic.gdx.graphics.Color
import com.badlogic.gdx.graphics.g2d.SpriteBatch
import com.badlogic.gdx.math.MathUtils
import com.badlogic.gdx.math.Vector2
import com.badlogic.gdx.physics.box2d.Filter
import ru.fredboy.cavedroid.common.api.SoundPlayer
import ru.fredboy.cavedroid.common.model.GameMode
import ru.fredboy.cavedroid.common.utils.TooltipManager
import ru.fredboy.cavedroid.common.utils.applyOrigin
import ru.fredboy.cavedroid.common.utils.drawSprite
import ru.fredboy.cavedroid.common.utils.meters
import ru.fredboy.cavedroid.domain.assets.repository.StepsSoundAssetsRepository
import ru.fredboy.cavedroid.domain.items.model.inventory.Inventory
import ru.fredboy.cavedroid.domain.items.model.inventory.InventoryItem
import ru.fredboy.cavedroid.domain.items.model.item.Item
import ru.fredboy.cavedroid.domain.items.model.mob.MobParams
import ru.fredboy.cavedroid.domain.items.usecase.GetFallbackItemUseCase
import ru.fredboy.cavedroid.domain.items.usecase.GetItemByKeyUseCase
import ru.fredboy.cavedroid.domain.world.model.PhysicsConstants
import ru.fredboy.cavedroid.entity.mob.abstraction.MobPhysicsFactory
import ru.fredboy.cavedroid.entity.mob.abstraction.MobWorldAdapter
import ru.fredboy.cavedroid.entity.mob.impl.PlayerMobBehavior

class Player(
    private val getFallbackItem: GetFallbackItemUseCase,
    private val tooltipManager: TooltipManager,
    params: MobParams,
    soundPlayer: SoundPlayer,
    stepsSoundAssetsRepository: StepsSoundAssetsRepository,

) : Mob(Direction.random(), params, PlayerMobBehavior(soundPlayer, stepsSoundAssetsRepository)) {

    var spawnPoint: Vector2? = null

    var inventory = Inventory(
        size = INVENTORY_SIZE,
        fallbackItem = getFallbackItem(),
    )
        set(value) {
            value.setOnItemAddedListener { item, slot ->
                if (slot == activeSlot) {
                    tooltipManager.showHotbarTooltip(item.item.params.name)
                }
            }
            field = value
        }

    var gameMode = GameMode.SURVIVAL

    var cursorX = 0f
    var cursorY = 0f

    var holdCursor = false
    var cursorToPlayer = Vector2()

    val selectedX get() = cursorX.toInt()
    val selectedY get() = cursorY.toInt()

    var controlMode = ControlMode.CURSOR
    var blockDamage = 0f

    var isHitting = false
    var isHittingWithDamage = false

    var hitAnim = 0f
    var hitAnimDelta = DEFAULT_ANIMATION_SPEED

    var headRotation = 0f

    private var _activeSlot = 0

    var isInBed = false

    var activeSlot
        get() = _activeSlot
        set(value) {
            if (value in 0..<HOTBAR_SIZE) {
                if (value != _activeSlot) {
                    tooltipManager.showHotbarTooltip(inventory.items[value].item.params.name)
                }
                _activeSlot = value
            }
        }

    val hotbarItems get() = inventory.items.subList(0, HOTBAR_SIZE)

    val activeItem get() = inventory.items[activeSlot]

    var sight: Light? = null
        set(value) {
            if (field != null) {
                field?.remove(true)
                field = null
            }

            field = value
        }

    override fun changeDir() = Unit

    override fun damage(damage: Int) {
        if (gameMode.isCreative()) {
            return
        }

        super.damage(damage)
    }

    override fun getDropItems(itemByKey: GetItemByKeyUseCase): List<InventoryItem> = inventory.items

    override fun draw(spriteBatch: SpriteBatch, x: Float, y: Float, delta: Float) {
        updateAnimation(delta)

        var backHandAnim: Float
        var frontHandAnim: Float

        val rightHandAnim = getRightHandAnim(delta)

        val backgroundTintColor = tintColor.cpy().sub(Color(0xAAAAAA shl 8))

        if (looksLeft()) {
            backHandAnim = rightHandAnim
            frontHandAnim = anim
        } else {
            backHandAnim = -anim
            frontHandAnim = -rightHandAnim
        }

        params.sprites.forEach { spriteData ->
            val sprite = spriteData.sprite

            sprite.setFlip(looksRight(), sprite.isFlipY)

            if (!isInBed) {
                sprite.applyOrigin(spriteData.origin)
            }

            if (spriteData.isBackground) {
                sprite.color = backgroundTintColor
            } else {
                sprite.color = tintColor
            }

            val animationValue = if (isInBed) {
                90f
            } else if (spriteData.isStatic) {
                0f
            } else if (spriteData.isHead) {
                headRotation
            } else if (spriteData.isHand && spriteData.isBackground) {
                backHandAnim
            } else if (spriteData.isHand) {
                frontHandAnim
            } else if (spriteData.isBackground) {
                anim
            } else {
                -anim
            }

            val spriteX = if (isInBed) {
                x - spriteData.offsetY
            } else {
                x + spriteData.offsetX
            }

            val spriteY = if (isInBed) {
                y - sprite.regionHeight.meters / 2 + height * 0.625f
            } else {
                y + spriteData.offsetY
            }

            if (spriteData.isHand && looksRight() && !spriteData.isBackground) {
                drawItem(spriteBatch, sprite.height, x, y, animationValue)
            }

            spriteBatch.drawSprite(sprite, spriteX, spriteY, animationValue)

            if (spriteData.isHand && spriteData.isBackground && looksLeft()) {
                drawItem(spriteBatch, sprite.height, x, y, -animationValue)
            }
        }
    }

    override fun dispose() {
        super.dispose()
        sight = null
    }

    fun respawn(spawnPoint: Vector2, mobPhysicsFactory: MobPhysicsFactory) {
        this.spawnPoint = spawnPoint
        isDead = false
        heal(maxHealth)
        breath = params.maxBreath
        spawn(spawnPoint.x, spawnPoint.y, mobPhysicsFactory)
        cursorX = position.x
        cursorY = position.y
    }

    fun initSight(mobWorldAdapter: MobWorldAdapter) {
        sight = PointLight(
            mobWorldAdapter.getRayHandler(),
            128,
            Color().apply { a = 0.5f },
            4f,
            position.x,
            position.y,
        ).apply {
            attachToBody(body)
            ignoreAttachedBody = true
            setContactFilter(
                Filter().apply {
                    categoryBits = PhysicsConstants.CATEGORY_OPAQUE
                },
            )
            isXray = true
        }
    }

    fun updateSight() {
        sight?.let { sight ->
            sight.setPosition(position.x, position.y)
            sight.direction = 0f
        }
    }

    fun startHitting(withDamage: Boolean = true) {
        if (isHitting) {
            return
        }

        isHitting = true
        isHittingWithDamage = withDamage
        hitAnim = HIT_ANIMATION_RANGE.endInclusive
        hitAnimDelta = DEFAULT_ANIMATION_SPEED
    }

    fun stopHitting() {
        blockDamage = 0f
        isHitting = false
    }

    fun hitMob(mob: Mob) {
        startHitting(false)
        val activeTool = activeItem.item as? Item.Tool
        if (activeTool != null) {
            decreaseCurrentItemCount()
        }
        val damage = 1 * (activeTool?.mobDamageMultiplier ?: 1f)
        mob.damage(damage.toInt())
        stopHitting()
    }

    fun decreaseCurrentItemCount(amount: Int = 1) {
        if (gameMode.isCreative()) {
            return
        }

        activeItem.subtract(amount)

        if (activeItem.amount <= 0) {
            setCurrentInventorySlotItem(getFallbackItem())
        }
    }

    fun setCurrentInventorySlotItem(item: Item) {
        inventory.items[activeSlot] = item.toInventoryItem()
    }

    private fun drawItem(spriteBatch: SpriteBatch, handLength: Float, x: Float, y: Float, handAnim: Float) {
        val item = activeItem.item.takeIf { !it.isNone() } ?: return
        val itemSprite = item.sprite
        val isSmallSprite = !item.isTool() || item.isShears()
        val originalWidth = itemSprite.width
        val originalHeight = itemSprite.height

        if (isSmallSprite) {
            itemSprite.setSize(SMALL_ITEM_SIZE, SMALL_ITEM_SIZE)
        }

        val spriteOrigin = item.params.inHandSpriteOrigin
        val handMultiplier = -direction.basis
        val xOffset = (-1 + direction.index) * itemSprite.width + SMALL_ITEM_SIZE / 2 +
            handMultiplier * itemSprite.width * spriteOrigin.x
        val yOffset = if (!isSmallSprite) -itemSprite.height / 2 else 0f

        val rotate = handAnim + HAND_ITEM_ANGLE_DEG

        if (item.isTool()) {
            itemSprite.rotate90(looksLeft())
        }

        val itemX = x + handLength * MathUtils.sin(handMultiplier * handAnim * MathUtils.degRad) + xOffset
        val itemY = y + handLength * MathUtils.cos(handMultiplier * handAnim * MathUtils.degRad) + yOffset

        if (looksLeft()) {
            itemSprite.setFlip(!item.isTool(), itemSprite.isFlipY)
            itemSprite.applyOrigin(spriteOrigin.getFlipped(flipX = true, flipY = false))
        } else {
            itemSprite.setFlip(item.isTool(), itemSprite.isFlipY)
            itemSprite.applyOrigin(spriteOrigin)
        }

        itemSprite.rotation = -handMultiplier * rotate
        itemSprite.setPosition(itemX, itemY)
        itemSprite.draw(spriteBatch)

        // dont forget to reset
        itemSprite.setFlip(false, itemSprite.isFlipY)
        itemSprite.rotation = 0f
        itemSprite.setOriginCenter()
        itemSprite.setSize(originalWidth, originalHeight)
        if (item.isTool()) {
            itemSprite.rotate90(looksRight())
        }
    }

    private fun getRightHandAnim(delta: Float): Float {
        hitAnim -= hitAnimDelta * delta

        if (hitAnim !in HIT_ANIMATION_RANGE) {
            if (isHitting) {
                hitAnim = MathUtils.clamp(hitAnim, HIT_ANIMATION_RANGE.start, HIT_ANIMATION_RANGE.endInclusive)
                hitAnimDelta = -hitAnimDelta
            } else {
                hitAnimDelta = DEFAULT_ANIMATION_SPEED
            }
        }

        if (!isHitting) {
            if (hitAnim < hitAnimDelta * delta) {
                hitAnim = 0f
                hitAnimDelta = 0f
                return -anim
            }
        }

        return hitAnim
    }

    enum class ControlMode {
        WALK,
        CURSOR,
    }

    companion object {
        const val HOTBAR_SIZE = 9
        const val INVENTORY_SIZE = 36

        private val HIT_ANIMATION_RANGE = 30f..90f

        private const val SMALL_ITEM_SIZE = .5f
        private const val HAND_ITEM_ANGLE_DEG = 30f
    }
}
